src/impl_methods.rs

```diff
@@ -1917,18 +1916,20 @@ where
         F: FnMut(&mut A, &B),
     {
         debug_assert_eq!(self.shape(), rhs.shape());
-        if let Some(self_s) = self.as_slice_mut() {
-            if let Some(rhs_s) = rhs.as_slice() {
-                let len = cmp::min(self_s.len(), rhs_s.len());
-                let s = &mut self_s[..len];
-                let r = &rhs_s[..len];
-                for i in 0..len {
-                    f(&mut s[i], &r[i]);
+
+        // Same shape and order should have same strides
+        if self.strides() == rhs.strides() {
```

Comment:
This is a little more restrictive than necessary. For axes with length <= 1, the stride doesn't matter. For example, if we have two arrays like this:

* shape: [1, 5], strides: [42, 1]
* shape: [1, 5], strides: [3, 1]

their memory layout is the same even though the strides for axis 0 are not equal (since the length of axis 0 is <= 1).

We can handle this case by adding a function like this to the `Dimension` trait (in `src/dimension/dimension_trait.rs`):

```rust
    /// Returns `true` iff `strides1` and `strides2` are equivalent for the
    /// shape `self`.
    ///
    /// The strides are equivalent if, for each axis with length > 1, the
    /// strides are equal.
    ///
    /// Note: Returns `false` if any of the ndims don't match.
    #[doc(hidden)]
    fn strides_equivalent(&self, strides1: &Self, strides2: &Self) -> bool {
        let shape_ndim = self.ndim();
        shape_ndim == strides1.ndim()
            && shape_ndim == strides2.ndim()
            && izip!(self.slice(), strides1.slice(), strides2.slice())
                .all(|(&d, &s1, &s2)| d <= 1 || s1 as isize == s2 as isize)
    }
```

and changing the check to

```rust
        if self.dim.strides_equivalent(&self.strides, &rhs.strides) {
```

I wouldn't expect this additional complication to have a significant performance impact (since I would expect most of the time to be in the iteration over the two arrays), but it's worth double-checking.